sdk/core/allocator/main.cc

```diff
@@ -216,20 +217,44 @@ namespace
 	                                 uint32_t                   epoch,
 	                                 LockGuard<decltype(lock)> &g,
 	                                 T                         &r = revoker)
-	    requires(!Revocation::SupportsInterruptNotification<T>)
+	    requires(!Revocation::SupportsInterruptNotification<T> &&
+	             Revocation::Revoker::IsAsynchronous)
 	{
 		// Yield while until a revocation pass has finished.
 		while (!revoker.has_revocation_finished_for_epoch(epoch))
 		{
 			// Release the lock before sleeping
 			g.unlock();
 			Timeout smallSleep{1};
-			if (thread_sleep(&smallSleep) < 0)
+			if ((thread_sleep(&smallSleep) < 0) ||
+			    !reacquire_lock(timeout, g, smallSleep.elapsed))
 			{
 				return false;
 			}
-			if (!reacquire_lock(timeout, g, smallSleep.elapsed))
+		}
+		return true;
+	}
+
+	template<typename T = Revocation::Revoker>
+	bool wait_for_background_revoker(Timeout                   *timeout,
```

Comment:
At a glance, this implementation looks to be the same as the above?

Reply:
Hmm, after the refactoring the without-interrupt one might now be equivalent to the synchronous one, yes.

Reply:
That still looks to be true, assuming I'm not missing something; can we cut this down to one rather than two?

Alternatively, should the one below more aggressively kick (that is, make incremental progress on) the revocation, rather than just one "unit" of work per `has_revocation_finished_for_epoch` call?